File: library/src/main/java/com/chaek/android/adapter/CommonSpanSizeLookup.java
package com.chaek.android.adapter;

import android.support.v7.widget.GridLayoutManager;


/**
 * @author Chaek
 */
public abstract class CommonSpanSizeLookup extends GridLayoutManager.SpanSizeLookup {
    private CommonAdapter commonAdapter;
    private int mSpanSize;

    public CommonSpanSizeLookup(CommonAdapter commonAdapter, int span) {
        this.mSpanSize = span;
        this.commonAdapter = commonAdapter;
    }

    @Override
    public int getSpanSize(int position) {
        if (commonAdapter.isHeadFootView(position)) {
            return mSpanSize;
        }
        return getCommonSpanSize(commonAdapter.getPosition(position), commonAdapter.getItemData(position));
    }

    /**
     * 获取list数据
     *
     * @param position CommonAdapter list的位置
     * @param listItem item数据
     * @return span
     */
    protected abstract int getCommonSpanSize(int position, Object listItem);

}

File: library/src/main/java/com/chaek/android/adapter/CommonSpanSizeLookup.kt
package com.chaek.android.adapter

import androidx.recyclerview.widget.GridLayoutManager


/**
 * @author Chaek
 */
abstract class CommonSpanSizeLookup(private val commonAdapter: CommonAdapter, private val mSpanSize: Int) : GridLayoutManager.SpanSizeLookup() {
    override fun getSpanSize(position: Int): Int {
        return if (commonAdapter.isHeadFootView(position)) {
            mSpanSize
        } else getCommonSpanSize(commonAdapter.getPosition(position), commonAdapter.getListItemData(position))
    }

    /**
     * 获取list数据
     *
     * @param position CommonAdapter list的位置
     * @param listItem item数据
     * @return span
     */
    protected abstract fun getCommonSpanSize(position: Int, listItem: Any?): Int

}